src/nodedef.cpp

```diff
@@ -213,10 +214,24 @@ void TileDef::serialize(std::ostream &os, u16 protocol_version) const
 		// Before f018737, TextureSource::getTextureAverageColor did not handle
 		// missing textures. "[png" can be used as base texture, but is not known
 		// on older clients. Hence use "blank.png" to avoid this problem.
-		if (!name.empty() && name[0] == '[')
-			os << serializeString16("blank.png^" + name);
-		else
+		// Note that modifiers such as [combine can be used as base textures
+		// and are compatible with older clients, so we could specifically single out [png:,
+		// but this would require the list here to be updated whenever a new base modifier
+		// is introduced. Instead, we whitelist the three cases that are supported by the 
+		// older clients and prefix anything else
```

Comment:
The comment doesn't mention why to whitelist [combine at all and generally I think this could be shorter:
```suggestion
		// Certain modifiers do not work identically when used with a base image, so we
		// need to exclude them from this workaround.
		// Note that "[png" could be singled out instead of these complex conditions,
		// but this was not done for forward-compatibility reasons.
```